include/pagmo/algorithms/gaco.hpp

```diff
@@ -0,0 +1,899 @@
+  * the GNU Lesser General Public License as published by the Free
+    Software Foundation; either version 3 of the License, or (at your
+    option) any later version.
+or
+  * the GNU General Public License as published by the Free Software
+    Foundation; either version 3 of the License, or (at your option) any
+    later version.
+or both in parallel, as here.
+The PaGMO library is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+You should have received copies of the GNU General Public License and the
+GNU Lesser General Public License along with the PaGMO library.  If not,
+see https://www.gnu.org/licenses/. */
+
+#ifndef PAGMO_ALGORITHMS_GACO_HPP
+#define PAGMO_ALGORITHMS_GACO_HPP
+
+#include <algorithm> // std::shuffle, std::transform
+#include <boost/math/constants/constants.hpp>
+#include <iomanip>
+#include <numeric> // std::iota, std::inner_product
+#include <random>
+#include <sstream> //std::osstringstream
+#include <string>
+#include <tuple>
+#include <utility> //std::swap
+#include <vector>
+
+#include <pagmo/algorithm.hpp> // needed for the cereal macro
+#include <pagmo/exceptions.hpp>
+#include <pagmo/io.hpp>
+#include <pagmo/population.hpp>
+#include <pagmo/problem.hpp>
+#include <pagmo/problems/decompose.hpp>
+#include <pagmo/rng.hpp>
+#include <pagmo/utils/generic.hpp>
+#include <pagmo/utils/generic.hpp>         // uniform_real_from_range
+#include <pagmo/utils/multi_objective.hpp> // crowding_distance, etc..
+
+namespace pagmo
+{
+/// Extended ACO
+/**
+ * ACO is inspired by the natural mechanism with which real ant colonies forage food.
+ * This algorithm has shown promising results in many trajectory optimization problems.
+ * The first appearance of the algorithm happened in Dr. Marco Dorigo's thesis, in 1992.
+ * ACO generates future generations of ants by using the a multi-kernel gaussian distribution
+ * based on three parameters (i.e., pheromone values) which are computed depending on the quality
+ * of each previous solution. The solutions are ranked through an oracle penalty method.
+ *
+ *
+ * The version implemented in pagmo can be applied to box-bounded multiple-objective optimization.
+ *
+ * See:  M. Schlueter, et al. (2009). Extended ant colony optimization for non-convex
+ * mixed integer non-linear programming. Computers & Operations Research.
+ */
+class g_aco
+{
+public:
+    /// Single entry of the log (gen, fevals, best_pen, best_fit, m_ker, worst_fit, m_oracle, dx, dp)
+    typedef std::tuple<unsigned, unsigned, double, double, unsigned, double, double, double, double> log_line_type;
+    /// The log
+    typedef std::vector<log_line_type> log_type;
+
+    /**
+     * Constructs the ACO user defined algorithm for single and multi-objective optimization.
+     *
+     * @param[in] gen Generations: number of generations to evolve.
```

Comment:
the order of these params must be the same as the order in the constructror args.